File: Editor/Tools/AnimationTools.cs
using System;
using UnityEngine;
using UnityEditor;
using UnityEditor.Animations;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Utility class for Animation tool operations
    /// </summary>
    internal static class AnimationToolUtils
    {
        /// <summary>
        /// Result of finding a GameObject
        /// </summary>
        public struct FindResult
        {
            public GameObject GameObject;
            public JObject Error;
        }

        /// <summary>
        /// Find a GameObject by instanceId or objectPath from parameters
        /// </summary>
        public static FindResult FindGameObject(JObject parameters)
        {
            int? instanceId = parameters["instanceId"]?.ToObject<int?>();
            string objectPath = parameters["objectPath"]?.ToObject<string>();

            GameObject gameObject = null;
            string identifierInfo = "";

            if (instanceId.HasValue)
            {
                gameObject = EditorUtility.InstanceIDToObject(instanceId.Value) as GameObject;
                identifierInfo = $"instance ID {instanceId.Value}";
            }
            else if (!string.IsNullOrEmpty(objectPath))
            {
                gameObject = GameObject.Find(objectPath);
                identifierInfo = $"path '{objectPath}'";
            }
            else
            {
                return new FindResult
                {
                    Error = McpUnitySocketHandler.CreateErrorResponse(
                        "Either 'instanceId' or 'objectPath' must be provided",
                        "validation_error"
                    )
                };
            }

            if (gameObject == null)
            {
                return new FindResult
                {
                    Error = McpUnitySocketHandler.CreateErrorResponse(
                        $"GameObject not found with {identifierInfo}",
                        "not_found_error"
                    )
                };
            }

            return new FindResult { GameObject = gameObject };
        }
    }

    /// <summary>
    /// Tool for getting Animator component information from a GameObject
    /// </summary>
    public class GetAnimatorInfoTool : McpToolBase
    {
        public GetAnimatorInfoTool()
        {
            Name = "get_animator_info";
            Description = "Gets detailed information about an Animator component including parameters, layers, and states.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            var findResult = AnimationToolUtils.FindGameObject(parameters);
            if (findResult.Error != null)
                return findResult.Error;

            GameObject go = findResult.GameObject;
            var animator = go.GetComponent<Animator>();
            if (animator == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"GameObject '{go.name}' does not have an Animator component",
                    "component_error"
                );
            }

            var controller = animator.runtimeAnimatorController;
            var animController = controller as AnimatorController;

            // Parameters
            var parametersArray = new JArray();
            if (animController != null)
            {
                foreach (var param in animController.parameters)
                {
                    parametersArray.Add(new JObject
                    {
                        ["name"] = param.name,
                        ["type"] = param.type.ToString(),
                        ["defaultFloat"] = param.defaultFloat,
                        ["defaultInt"] = param.defaultInt,
                        ["defaultBool"] = param.defaultBool
                    });
                }
            }

            // Layers and States
            var layers = new JArray();
            if (animController != null)
            {
                foreach (var layer in animController.layers)
                {
                    var states = new JArray();
                    foreach (var childState in layer.stateMachine.states)
                    {
                        states.Add(new JObject
                        {
                            ["name"] = childState.state.name,
                            ["speed"] = childState.state.speed,
                            ["motion"] = childState.state.motion?.name ?? "None"
                        });
                    }
                    layers.Add(new JObject
                    {
                        ["name"] = layer.name,
                        ["defaultWeight"] = layer.defaultWeight,
                        ["states"] = states
                    });
                }
            }

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Animator info for '{go.name}'",
                ["parameters"] = parametersArray,
                ["layers"] = layers,
                ["controllerName"] = controller?.name ?? "None"
            };
        }
    }

    /// <summary>
    /// Tool for setting Animator parameters on a GameObject
    /// </summary>
    public class SetAnimatorParameterTool : McpToolBase
    {
        public SetAnimatorParameterTool()
        {
            Name = "set_animator_parameter";
            Description = "Sets an Animator parameter value. Supports Float, Int, Bool, and Trigger types with auto-detection.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            var findResult = AnimationToolUtils.FindGameObject(parameters);
            if (findResult.Error != null)
                return findResult.Error;

            GameObject go = findResult.GameObject;
            var animator = go.GetComponent<Animator>();
            if (animator == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"GameObject '{go.name}' does not have an Animator component",
                    "component_error"
                );
            }

            string parameterName = parameters["parameterName"]?.ToObject<string>();
            if (string.IsNullOrEmpty(parameterName))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'parameterName' not provided",
                    "validation_error"
                );
            }

            JToken valueToken = parameters["value"];
            string parameterTypeStr = parameters["parameterType"]?.ToObject<string>();

            // Auto-detect parameter type from AnimatorController
            AnimatorControllerParameterType paramType;
            var controller = animator.runtimeAnimatorController as AnimatorController;
            if (controller != null && string.IsNullOrEmpty(parameterTypeStr))
            {
                bool found = false;
                foreach (var param in controller.parameters)
                {
                    if (param.name == parameterName)
                    {
                        paramType = param.type;
                        found = true;
                        return ApplyParameter(animator, parameterName, paramType, valueToken, go.name);
                    }
                }
                if (!found)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Parameter '{parameterName}' not found in Animator controller",
                        "not_found_error"
                    );
                }
            }

            // Use explicit parameterType if provided
            if (!string.IsNullOrEmpty(parameterTypeStr))
            {
                if (!Enum.TryParse<AnimatorControllerParameterType>(parameterTypeStr, true, out paramType))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Invalid parameter type '{parameterTypeStr}'. Must be Float, Int, Bool, or Trigger",
                        "validation_error"
                    );
                }
                return ApplyParameter(animator, parameterName, paramType, valueToken, go.name);
            }

            return McpUnitySocketHandler.CreateErrorResponse(
                "Could not determine parameter type. Provide 'parameterType' or ensure the Animator has an AnimatorController",
                "validation_error"
            );
        }

        private JObject ApplyParameter(Animator animator, string parameterName, AnimatorControllerParameterType paramType, JToken valueToken, string goName)
        {
            Undo.RecordObject(animator, $"Set Animator Parameter {parameterName}");

            switch (paramType)
            {
                case AnimatorControllerParameterType.Float:
                    float floatVal = valueToken?.ToObject<float>() ?? 0f;
                    animator.SetFloat(parameterName, floatVal);
                    break;
                case AnimatorControllerParameterType.Int:
                    int intVal = valueToken?.ToObject<int>() ?? 0;
                    animator.SetInteger(parameterName, intVal);
                    break;
                case AnimatorControllerParameterType.Bool:
                    bool boolVal = valueToken?.ToObject<bool>() ?? false;
                    animator.SetBool(parameterName, boolVal);
                    break;
                case AnimatorControllerParameterType.Trigger:
                    animator.SetTrigger(parameterName);
                    break;
            }

            EditorUtility.SetDirty(animator);

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Set parameter '{parameterName}' ({paramType}) on '{goName}'"
            };
        }
    }
}

File: Server~/src/tools/animationTools.ts
import * as z from 'zod';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// ============================================================================
// get_animator_info Tool
// ============================================================================

const getAnimatorInfoToolName = 'get_animator_info';
const getAnimatorInfoToolDescription = 'Gets detailed information about an Animator component including parameters, layers, and states.';
const getAnimatorInfoParamsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy (alternative to instanceId)')
});

/**
 * Registers the get_animator_info tool with the MCP server
 */
export function registerGetAnimatorInfoTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${getAnimatorInfoToolName}`);

  server.tool(
    getAnimatorInfoToolName,
    getAnimatorInfoToolDescription,
    getAnimatorInfoParamsSchema.shape,
    async (params: z.infer<typeof getAnimatorInfoParamsSchema>) => {
      try {
        logger.info(`Executing tool: ${getAnimatorInfoToolName}`, params);
        const result = await getAnimatorInfoHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${getAnimatorInfoToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${getAnimatorInfoToolName}`, error);
        throw error;
      }
    }
  );
}

async function getAnimatorInfoHandler(mcpUnity: McpUnity, params: z.infer<typeof getAnimatorInfoParamsSchema>): Promise<CallToolResult> {
  validateGameObjectIdentifier(params);

  const response = await mcpUnity.sendRequest({
    method: getAnimatorInfoToolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to get animator info'
    );
  }

  // Format the response
  let text = `Animator Info for controller: ${response.controllerName}\n\n`;

  text += `Parameters:\n`;
  if (response.parameters && Array.isArray(response.parameters)) {
    for (const param of response.parameters) {
      text += `  - ${param.name} (${param.type})`;
      if (param.type === 'Float') text += ` default: ${param.defaultFloat}`;
      else if (param.type === 'Int') text += ` default: ${param.defaultInt}`;
      else if (param.type === 'Bool') text += ` default: ${param.defaultBool}`;
      text += '\n';
    }
  }

  text += `\nLayers:\n`;
  if (response.layers && Array.isArray(response.layers)) {
    for (const layer of response.layers) {
      text += `  ${layer.name} (weight: ${layer.defaultWeight})\n`;
      if (layer.states && Array.isArray(layer.states)) {
        for (const state of layer.states) {
          text += `    - ${state.name} (speed: ${state.speed}, motion: ${state.motion})\n`;
        }
      }
    }
  }

  return {
    content: [{
      type: 'text',
      text: text
    }],
    data: {
      controllerName: response.controllerName,
      parameters: response.parameters,
      layers: response.layers
    }
  };
}

// ============================================================================
// set_animator_parameter Tool
// ============================================================================

const setAnimatorParameterToolName = 'set_animator_parameter';
const setAnimatorParameterToolDescription = 'Sets an Animator parameter value. Supports Float, Int, Bool, and Trigger types with auto-detection from the AnimatorController.';
const setAnimatorParameterParamsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy (alternative to instanceId)'),
  parameterName: z.string().describe('The name of the Animator parameter to set'),
  value: z.any().optional().describe('The value to set (number for Float/Int, boolean for Bool, not needed for Trigger)'),
  parameterType: z.enum(['Float', 'Int', 'Bool', 'Trigger']).optional().describe('The parameter type. Auto-detected from the AnimatorController if not specified')
});

/**
 * Registers the set_animator_parameter tool with the MCP server
 */
export function registerSetAnimatorParameterTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${setAnimatorParameterToolName}`);

  server.tool(
    setAnimatorParameterToolName,
    setAnimatorParameterToolDescription,
    setAnimatorParameterParamsSchema.shape,
    async (params: z.infer<typeof setAnimatorParameterParamsSchema>) => {
      try {
        logger.info(`Executing tool: ${setAnimatorParameterToolName}`, params);
        const result = await setAnimatorParameterHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${setAnimatorParameterToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${setAnimatorParameterToolName}`, error);
        throw error;
      }
    }
  );
}

async function setAnimatorParameterHandler(mcpUnity: McpUnity, params: z.infer<typeof setAnimatorParameterParamsSchema>): Promise<CallToolResult> {
  validateGameObjectIdentifier(params);

  if (!params.parameterName || params.parameterName.trim() === '') {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'parameterName' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: setAnimatorParameterToolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath,
      parameterName: params.parameterName,
      value: params.value,
      parameterType: params.parameterType
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to set animator parameter'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || 'Animator parameter set successfully'
    }]
  };
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Validates that either instanceId or objectPath is provided
 */
function validateGameObjectIdentifier(params: { instanceId?: number; objectPath?: string }) {
  if ((params.instanceId === undefined || params.instanceId === null) &&
      (!params.objectPath || params.objectPath.trim() === '')) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Either 'instanceId' or 'objectPath' must be provided"
    );
  }
}

/**
 * Registers all animation tools with the MCP server
 */
export function registerAnimationTools(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  registerGetAnimatorInfoTool(server, mcpUnity, logger);
  registerSetAnimatorParameterTool(server, mcpUnity, logger);
}
